eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs

```diff
@@ -0,0 +1,214 @@
+
+instance Era era => EncCBOR (ConwayDelegPredFailure era) where
+  encCBOR =
+    encode . \case
+      IncorrectDepositDELEG mCoin ->
+        Sum (IncorrectDepositDELEG @era) 1
+          !> To mCoin
+      StakeKeyAlreadyRegisteredDELEG stakeCred ->
+        Sum (StakeKeyAlreadyRegisteredDELEG @era) 2
+          !> To stakeCred
+      StakeKeyNotRegisteredDELEG stakeCred ->
+        Sum (StakeKeyNotRegisteredDELEG @era) 3
+          !> To stakeCred
+      StakeKeyHasNonZeroAccountBalanceDELEG mCoin ->
+        Sum (StakeKeyHasNonZeroAccountBalanceDELEG @era) 4
+          !> To mCoin
+      DRepAlreadyRegisteredForStakeKeyDELEG stakeCred ->
+        Sum (DRepAlreadyRegisteredForStakeKeyDELEG @era) 5
+          !> To stakeCred
+      WrongCertificateTypeDELEG ->
+        Sum (WrongCertificateTypeDELEG @era) 6
+
+instance Era era => DecCBOR (ConwayDelegPredFailure era) where
+  decCBOR = decode $ Summands "ConwayDelegPredFailure" $ \case
+    1 -> SumD IncorrectDepositDELEG <! From
+    2 -> SumD StakeKeyAlreadyRegisteredDELEG <! From
+    3 -> SumD StakeKeyNotRegisteredDELEG <! From
+    4 -> SumD StakeKeyHasNonZeroAccountBalanceDELEG <! From
+    5 -> SumD DRepAlreadyRegisteredForStakeKeyDELEG <! From
+    6 -> SumD WrongCertificateTypeDELEG
+    n -> Invalid n
+
+newtype ConwayDelegEvent era = DelegEvent (Event (EraRule "DELEG" era))
+
+instance
+  ( EraPParams era
+  , State (EraRule "DELEG" era) ~ DState era
+  , Signal (EraRule "DELEG" era) ~ ConwayDelegCert (EraCrypto era)
+  , Environment (EraRule "DELEG" era) ~ DelegEnv era
+  , EraRule "DELEG" era ~ ConwayDELEG era
+  ) =>
+  STS (ConwayDELEG era)
+  where
+  type State (ConwayDELEG era) = DState era
+  type Signal (ConwayDELEG era) = ConwayDelegCert (EraCrypto era)
+  type Environment (ConwayDELEG era) = DelegEnv era
+  type BaseM (ConwayDELEG era) = ShelleyBase
+  type PredicateFailure (ConwayDELEG era) = ConwayDelegPredFailure era
+  type Event (ConwayDELEG era) = ConwayDelegEvent era
+
+  transitionRules = [conwayDelegTransition @era]
+
+conwayDelegTransition :: forall era. EraPParams era => TransitionRule (ConwayDELEG era)
+conwayDelegTransition = do
+  TRC
+    ( DelegEnv _slot _ptr _acnt pp
+      , dState@DState {dsUnified}
+      , c
+      ) <-
+    judgmentContext
+  let pd = pp ^. ppKeyDepositL
+  case c of
+    ConwayRegCert stakeCred sMayDeposit -> do
+      checkDepositAgainstPParams pd sMayDeposit
+      checkStakeKeyNotAlreadyRegistered stakeCred dsUnified
+      pure $ dState {dsUnified = acceptDepositForStakeKey stakeCred dsUnified pd}
+    ConwayUnRegCert stakeCred sMayDeposit -> do
+      checkStakeKeyIsAlreadyRegistered stakeCred dsUnified
+      checkStakeKeyHasZeroBalance stakeCred dsUnified
+      unless (sMayDeposit == SNothing) $ checkDepositAgainstPaidDeposit stakeCred dsUnified sMayDeposit
+      let umRDRemoved = Set.singleton stakeCred UM.⋪ UM.RewDepUView dsUnified
+          umSPoolRemoved = Set.singleton stakeCred UM.⋪ UM.SPoolUView umRDRemoved
```

Comment:
Is it possible to remove the whole UMap entry? We shouldn't release the deposit unless all of the memory allocated by the delegation is freed.

Reply:
at the moment we don't remove the DRep delegation

Reply:
at the very least we should remove the DRep deposits as well

Reply:
I agree, it would be best to add a specialized function in the UMap module for removing credentials. Also in the name of performance and future compatibility:

```haskell
delCredUMap :: Crypto c => Credential Staking' c -> UMap c -> UMap c
```